core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java

```diff
@@ -1265,6 +1275,66 @@ public static <T> void toHttp1Headers(HttpHeaders armeriaHeaders, T output,
         }
     }
 
+    /**
+     * Attempts to parse this URI's authority component and return {@link URI}.
+     * The authority part may have one of the following formats (The userinfo part will be ignored.):
+     * <ul>
+     *   <li>{@code "<host>:<port>"} for a host endpoint </li>
+     *   <li>{@code "<host>"}, {@code "<host>:"} for a host endpoint with no port number specified</li>
+     * </ul>
+     * An IPv4 or IPv6 address can be specified in lieu of a host name, e.g. {@code "127.0.0.1:8080"} and
+     * {@code "[::1]:8080"}.
+     *
+     * @throws IllegalArgumentException if {@code authority} do not comply with RFC 2396
+     */
+    public static URI normalizeAuthority(String authority) {
```

Comment:
In order to handle the domain socket case, I guess we have two options:

1. Change the return type of `normalizeAuthority` to a type which encompasses both domain sockets and non-domain sockets. If this is under the `*internal*` package, users wouldn't prefer to access this class.

```java
static class NormalizedAuthority { // TODO: maybe a better name
  String scheme;
  String authority;
  int port;
}
```

```java
if (isDomainSocketAuthority(rawAuthority)) {
  return NormalizedAuthority(rawAuthority);
}
// do normal validation
```

2. Alternatively, we could delegate domain socket calls to the caller (invoker). This means the following code would be used in `Endpoint#parse`, `DefaultClientRequestContext#autoFillSchemeAuthorityAndOrigin`, etc..

```java
if (isDomainSocketAuthority(rawAuthority)) {
  normalized = rawAuthority
} else {
  normalized = normalizeAuthority(rawAuthority)
}
```

I think the former is more maintainable at the cost of an extra allocation. Let me know what you think though

Reply:
OMG, I didn't consider the domain socket case!!
I also think the former is easy to use and maintain. so,  I'll make a class under *internal/common* package that's responsible for authority normalization!!

Reply:
I made `SchemeAndAuthority` which is in charge of authority normalization!